File: src/commands/crud.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { getContext, handleError, extractAttributes, filterColumns } from '../context';
import { DaptinQueryParameters } from '../client/types';

export function registerCrudCommands(program: Command): void {
  // --- LIST ---
  program
    .command('list <table>')
    .description('List records from a table')
    .option('--page-size <n>', 'Records per page', '10')
    .option('--page <n>', 'Page number (1-indexed)', '1')
    .option('--filter <json>', 'Filter conditions (JSON array)')
    .option('--sort <fields>', 'Sort fields (prefix with - for desc)')
    .option('--columns <cols>', 'Comma-separated columns to display')
    .option('--include <rels>', 'Include relationships')
    .option('--fields <fields>', 'Fields to request from server')
    .action(async (table: string, options: any, cmd: Command) => {
      try {
        const { client, renderer } = getContext(cmd);
        const params: DaptinQueryParameters = {
          'page[size]': parseInt(options.pageSize, 10),
          'page[number]': parseInt(options.page, 10),
        };

        if (options.filter) params['query'] = options.filter;
        if (options.sort) params['sort'] = options.sort;
        if (options.include) params['included_relations'] = options.include;
        if (options.fields) params['fields'] = options.fields;

        const response = await client.findAll(table, params);
        let data = extractAttributes(response.data);

        if (data.length === 0) {
          console.log(chalk.yellow('No records found.'));
          return;
        }

        const columns = options.columns
          ? options.columns.split(',').map((c: string) => c.trim())
          : undefined;

        if (columns) {
          data = filterColumns(data, columns);
        }

        renderer.renderArray(data, columns);

        // Show pagination info
        if (response.links) {
          const links = response.links as any;
          const total = links.total || data.length;
          const currentPage = links.current_page || parseInt(options.page, 10);
          const lastPage = links.last_page || 1;
          console.log(
            chalk.gray(`\nPage ${currentPage}/${lastPage} (${total} total records)`)
          );
        }
      } catch (error) {
        handleError(error);
      }
    });

  // --- GET ---
  program
    .command('get <table> <referenceId>')
    .description('Get a single record by reference_id')
    .option('--columns <cols>', 'Comma-separated columns to display')
    .option('--include <rels>', 'Include relationships')
    .action(async (table: string, referenceId: string, options: any, cmd: Command) => {
      try {
        const { client, renderer } = getContext(cmd);
        const params: DaptinQueryParameters = {};

        if (options.include) params['included_relations'] = options.include;

        const response = await client.findOne(table, referenceId, params);

        if (!response.data) {
          console.log(chalk.yellow('Record not found.'));
          return;
        }

        const attrs = response.data.attributes || {};
        const columns = options.columns
          ? options.columns.split(',').map((c: string) => c.trim())
          : undefined;

        renderer.renderObject(attrs, columns);
      } catch (error) {
        handleError(error);
      }
    });

  // --- CREATE ---
  program
    .command('create <table>')
    .description('Create a new record')
    .requiredOption('--data <json>', 'JSON object with record attributes')
    .action(async (table: string, options: any, cmd: Command) => {
      try {
        const { client, renderer } = getContext(cmd);
        let attributes: Record<string, any>;

        try {
          attributes = JSON.parse(options.data);
        } catch {
          console.error(chalk.red('Invalid JSON in --data flag.'));
          process.exit(1);
        }

        const response = await client.create(table, attributes);
        const attrs = response.data?.attributes || {};

        console.log(chalk.green('Record created successfully.'));
        renderer.renderObject(attrs);
      } catch (error) {
        handleError(error);
      }
    });

  // --- UPDATE ---
  program
    .command('update <table> <referenceId>')
    .description('Update an existing record')
    .requiredOption('--data <json>', 'JSON object with attributes to update')
    .action(async (table: string, referenceId: string, options: any, cmd: Command) => {
      try {
        const { client, renderer } = getContext(cmd);
        let attributes: Record<string, any>;

        try {
          attributes = JSON.parse(options.data);
        } catch {
          console.error(chalk.red('Invalid JSON in --data flag.'));
          process.exit(1);
        }

        const response = await client.update(table, referenceId, attributes);
        const attrs = response.data?.attributes || {};

        console.log(chalk.green('Record updated successfully.'));
        renderer.renderObject(attrs);
      } catch (error) {
        handleError(error);
      }
    });

  // --- DELETE ---
  program
    .command('delete <table> <referenceId>')
    .description('Delete a record by reference_id')
    .action(async (table: string, referenceId: string, _options: any, cmd: Command) => {
      try {
        const { client } = getContext(cmd);
        await client.deleteOne(table, referenceId);
        console.log(chalk.green(`Record ${referenceId} deleted from "${table}".`));
      } catch (error) {
        handleError(error);
      }
    });

  // --- RELATION ---
  program
    .command('relation <table> <referenceId> <relation>')
    .description('Query related records through a relationship')
    .option('--page-size <n>', 'Records per page', '10')
    .option('--page <n>', 'Page number', '1')
    .option('--columns <cols>', 'Comma-separated columns to display')
    .action(
      async (
        table: string,
        referenceId: string,
        relation: string,
        options: any,
        cmd: Command
      ) => {
        try {
          const { client, renderer } = getContext(cmd);
          const params: DaptinQueryParameters = {
            'page[size]': parseInt(options.pageSize, 10),
            'page[number]': parseInt(options.page, 10),
          };

          const response = await client.getRelation(table, referenceId, relation, params);

          if (Array.isArray(response.data)) {
            const data = extractAttributes(response.data);
            if (data.length === 0) {
              console.log(chalk.yellow('No related records found.'));
              return;
            }
            const columns = options.columns
              ? options.columns.split(',').map((c: string) => c.trim())
              : undefined;
            renderer.renderArray(columns ? filterColumns(data, columns) : data, columns);
          } else if (response.data) {
            const attrs = (response.data as any).attributes || {};
            renderer.renderObject(attrs);
          } else {
            console.log(chalk.yellow('No related records found.'));
          }
        } catch (error) {
          handleError(error);
        }
      }
    );
}

File: cmd/crud.go
package cmd

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"

	"github.com/daptin/daptin-cli/client"
	"github.com/daptin/daptin-cli/render"
	daptinClient "github.com/daptin/daptin-go-client"
	"github.com/urfave/cli/v2"
)

func listCommand(appCtx *AppContext) *cli.Command {
	return &cli.Command{
		Name:      "list",
		Usage:     "List rows of an entity",
		ArgsUsage: "<entity>",
		UsageText: `daptin list <entity> [flags]
   daptin list usergroup --filter name=administrators --columns name,reference_id
   daptin list world --filter "table_name like %doc%" --page-size 50
   daptin list document --sort -created_at`,
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name:  "columns",
				Usage: "Comma-separated column names to show",
			},
			&cli.IntFlag{
				Name:  "page-size",
				Usage: "Number of items per page",
				Value: 10,
			},
			&cli.IntFlag{
				Name:  "page",
				Usage: "Page number",
				Value: 1,
			},
			&cli.StringFlag{
				Name:  "sort",
				Usage: "Sort column (prefix - for descending)",
			},
			&cli.StringFlag{
				Name:  "filter",
				Usage: "Filter expression, e.g. name=value or \"name is value\"",
			},
			&cli.StringFlag{
				Name:  "include",
				Usage: "Comma-separated relation names to include",
			},
		},
		Action: func(c *cli.Context) error {
			entityName := c.Args().Get(0)
			if entityName == "" {
				return fmt.Errorf("entity name required")
			}
			slog.Info("list", "entity", entityName, "page", c.Int("page"), "page_size", c.Int("page-size"))

			params := daptinClient.DaptinQueryParameters{
				"page[size]":   c.Int("page-size"),
				"page[number]": c.Int("page"),
			}
			if s := c.String("sort"); s != "" {
				params["sort"] = s
			}
			if f := c.String("filter"); f != "" {
				clauses, err := ParseFilter(f)
				if err != nil {
					return err
				}
				params["query"] = FilterToJSON(clauses)
			}
			if inc := c.String("include"); inc != "" {
				params["included_relations"] = inc
			}

			result, err := appCtx.Client.FindAll(entityName, params)
			if err != nil {
				return err
			}
			if len(result) == 0 {
				fmt.Println("No rows found")
				return nil
			}

			rows := client.MapArray(result, "attributes")
			slog.Debug("list results", "count", len(rows))
			if appCtx.Quiet {
				return printRefs(rows)
			}
			if cols := c.String("columns"); cols != "" {
				colList := strings.Split(cols, ",")
				slog.Debug("filtering columns", "columns", colList)
				rows = render.FilterColumns(rows, colList)
			}
			return appCtx.Renderer.RenderArray(rows)
		},
	}
}

func getCommand(appCtx *AppContext) *cli.Command {
	return &cli.Command{
		Name:      "get",
		Usage:     "Get a single row by reference_id",
		ArgsUsage: "<entity> <reference_id>",
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name:  "columns",
				Usage: "Comma-separated column names to show",
			},
		},
		Action: func(c *cli.Context) error {
			entityName := c.Args().Get(0)
			referenceId := c.Args().Get(1)
			if entityName == "" || referenceId == "" {
				return fmt.Errorf("usage: get <entity> <reference_id>")
			}
			slog.Info("get", "entity", entityName, "reference_id", referenceId)

			result, err := appCtx.Client.FindOne(entityName, referenceId, nil)
			if err != nil {
				return err
			}

			row, ok := result["attributes"].(map[string]interface{})
			if !ok {
				fmt.Println("No data found")
				return nil
			}

			if appCtx.Quiet {
				return printRef(row)
			}
			if cols := c.String("columns"); cols != "" {
				row = render.IncludeColumns(row, strings.Split(cols, ","))
			}
			return appCtx.Renderer.RenderObject(row)
		},
	}
}

func createCommand(appCtx *AppContext) *cli.Command {
	return &cli.Command{
		Name:      "create",
		Usage:     "Create a new row",
		ArgsUsage: "<entity> [key=val ...] or <entity> <json>",
		Action: func(c *cli.Context) error {
			entityName := c.Args().Get(0)
			if entityName == "" {
				return fmt.Errorf("entity name required")
			}
			slog.Info("create", "entity", entityName)

			attrs, err := parseAttributes(c.Args().Slice()[1:])
			if err != nil {
				return err
			}

			body := daptinClient.JsonApiObject{
				"data": map[string]interface{}{
					"type":       entityName,
					"attributes": attrs,
				},
			}

			result, err := appCtx.Client.Create(entityName, body)
			if err != nil {
				return err
			}

			data, _ := result["attributes"].(map[string]interface{})
			if data == nil {
				data = result
			}
			if appCtx.Quiet {
				return printRef(data)
			}
			return appCtx.Renderer.RenderObject(data)
		},
	}
}

func updateCommand(appCtx *AppContext) *cli.Command {
	return &cli.Command{
		Name:      "update",
		Usage:     "Update a row",
		ArgsUsage: "<entity> <reference_id> [key=val ...] or <entity> <reference_id> <json>",
		Action: func(c *cli.Context) error {
			entityName := c.Args().Get(0)
			referenceId := c.Args().Get(1)
			if entityName == "" || referenceId == "" {
				return fmt.Errorf("usage: update <entity> <reference_id> [key=val ...]")
			}
			slog.Info("update", "entity", entityName, "reference_id", referenceId)

			attrs, err := parseAttributes(c.Args().Slice()[2:])
			if err != nil {
				return err
			}

			body := daptinClient.JsonApiObject{
				"data": map[string]interface{}{
					"type":       entityName,
					"id":         referenceId,
					"attributes": attrs,
				},
			}

			result, err := appCtx.Client.Update(entityName, referenceId, body)
			if err != nil {
				return err
			}

			data, _ := result["attributes"].(map[string]interface{})
			if data == nil {
				data = result
			}
			if appCtx.Quiet {
				return printRef(data)
			}
			return appCtx.Renderer.RenderObject(data)
		},
	}
}

func deleteCommand(appCtx *AppContext) *cli.Command {
	return &cli.Command{
		Name:      "delete",
		Usage:     "Delete a row",
		ArgsUsage: "<entity> <reference_id>",
		Action: func(c *cli.Context) error {
			entityName := c.Args().Get(0)
			referenceId := c.Args().Get(1)
			if entityName == "" || referenceId == "" {
				return fmt.Errorf("usage: delete <entity> <reference_id>")
			}
			slog.Info("delete", "entity", entityName, "reference_id", referenceId)

			err := appCtx.Client.Delete(entityName, referenceId)
			if err != nil {
				return err
			}
			fmt.Println("Deleted")
			return nil
		},
	}
}

func relatedCommand(appCtx *AppContext) *cli.Command {
	return &cli.Command{
		Name:      "related",
		Usage:     "Get related rows via a relationship",
		ArgsUsage: "<entity> <reference_id> <relation>",
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name:  "columns",
				Usage: "Comma-separated column names to show",
			},
		},
		Action: func(c *cli.Context) error {
			entityName := c.Args().Get(0)
			referenceId := c.Args().Get(1)
			relation := c.Args().Get(2)
			if entityName == "" || referenceId == "" || relation == "" {
				return fmt.Errorf("usage: related <entity> <reference_id> <relation>")
			}
			slog.Info("related", "entity", entityName, "reference_id", referenceId, "relation", relation)

			result, err := appCtx.Client.FindRelated(entityName, referenceId, relation, nil)
			if err != nil {
				return err
			}

			rows := client.MapArray(result, "attributes")
			if appCtx.Quiet {
				return printRefs(rows)
			}
			if cols := c.String("columns"); cols != "" {
				rows = render.FilterColumns(rows, strings.Split(cols, ","))
			}
			return appCtx.Renderer.RenderArray(rows)
		},
	}
}

// printRef outputs only the reference_id from a single row.
func printRef(row map[string]interface{}) error {
	if ref, ok := row["reference_id"].(string); ok {
		fmt.Println(ref)
	}
	return nil
}

// printRefs outputs one reference_id per line from a list of rows.
func printRefs(rows []map[string]interface{}) error {
	for _, row := range rows {
		if ref, ok := row["reference_id"].(string); ok {
			fmt.Println(ref)
		}
	}
	return nil
}

// parseAttributes parses [key=val ...] args or a single JSON string into a map.
func parseAttributes(args []string) (map[string]interface{}, error) {
	if len(args) == 0 {
		return map[string]interface{}{}, nil
	}

	// If the first arg looks like JSON, parse it
	first := args[0]
	if strings.HasPrefix(first, "{") {
		joined := strings.Join(args, " ")
		var result map[string]interface{}
		if err := json.Unmarshal([]byte(joined), &result); err != nil {
			return nil, fmt.Errorf("invalid JSON: %w", err)
		}
		return result, nil
	}

	// Otherwise parse key=val pairs
	result := make(map[string]interface{}, len(args))
	for _, arg := range args {
		parts := strings.SplitN(arg, "=", 2)
		if len(parts) != 2 {
			return nil, fmt.Errorf("invalid argument %q, expected key=value", arg)
		}
		result[parts[0]] = parts[1]
	}
	return result, nil
}
